File: schema.py
from typing import List, Optional, Dict, Any
from pydantic import BaseModel
from datetime import datetime


class ContentSection(BaseModel):
    type: str  # "section", "note", "code", "table", "link", "video"
    heading: Optional[str] = None
    text: Optional[str] = None
    code: Optional[str] = None
    language: Optional[str] = None
    note: Optional[str] = None
    table: Optional[Dict[str, Any]] = None
    url: Optional[str] = None


class BlogPost(BaseModel):
    id: str
    slug: str
    title: str
    subtitle: str
    excerpt: str
    content: List[ContentSection]
    author: str
    publishedDate: datetime
    readTime: int
    tags: List[str]
    image: str
    category: str
    featured: bool
    views: int
    likes: int

File: schema.js
import { z } from "zod";

const TextSection = z.object({
  title: z.string(),
  type: z.literal("text"),
  content: z.string(),
});

const BulletsSection = z.object({
  title: z.string(),
  type: z.literal("bullets"),
  items: z.array(z.string()),
});

const ImageSection = z.object({
  title: z.string(),
  type: z.literal("image"),
  url: z.string(),
  alt: z.string().optional(),
  caption: z.string().optional(),
});

const CodeSection = z.object({
  title: z.string(),
  type: z.literal("code"),
  language: z.string(),
  content: z.string(),
});

const TableSection = z.object({
  title: z.string(),
  type: z.literal("table"),
  headers: z.array(z.string()),
  rows: z.array(z.array(z.string())),
});

const NoteSection = z.object({
  title: z.string(),
  type: z.literal("note"),
  content: z.string(),
});

const YouTubeSection = z.object({
  title: z.string(),
  type: z.literal("youtube"),
  videoId: z.string(),
  description: z.string().optional(),
});

const LinksSection = z.object({
  title: z.string(),
  type: z.literal("links"),
  links: z.array(
    z.object({
      text: z.string(),
      url: z.string(),
      description: z.string().optional(),
    })
  ),
});

const Section = z.union([
  TextSection,
  BulletsSection,
  ImageSection,
  CodeSection,
  TableSection,
  NoteSection,
  YouTubeSection,
  LinksSection,
]);

const Content = z.object({
  introduction: z.string(),
  sections: z.array(Section),
  conclusion: z.string(),
});

const Blog = z.object({
  id: z.number(),
  slug: z.string(),
  title: z.string(),
  subtitle: z.string(),
  excerpt: z.string(),
  content: Content,
  author: z.string(),
  publishedDate: z.string(),
  readTime: z.string(),
  tags: z.array(z.string()),
  image: z.string(),
  category: z.string(),
  featured: z.boolean(),
  views: z.number(),
  likes: z.number(),
});

export const BlogsSchema = z.object({
  blogs: z.array(Blog),
});

